src/core/expression/qgsexpressionnodeimpl.cpp

```diff
@@ -694,6 +694,122 @@ QgsExpressionNode::NodeType QgsExpressionNodeBinaryOperator::nodeType() const
+          if ( dynamic_cast<QgsExpressionNodeColumnRef *>( op->opLeft() ) && dynamic_cast<QgsExpressionNodeLiteral *>( op->opRight() ) )
+          {
+            const QString fieldName = op->opLeft()->dump();
+            if ( !orValuesMap.contains( fieldName ) )
+            {
+              orFieldNames.append( fieldName );
+              orValuesMap.insert( fieldName, QgsExpressionNode::NodeList() );
+            }
+            orValuesMap[fieldName].append( op->opRight()->clone() );
+            return true;
+          }
+          return false;
+        }
+
+        if ( visitOrNodes( op->opLeft() ) && visitOrNodes( op->opRight() ) )
+        {
+          return true;
+        }
+
+      }
+      else if ( QgsExpressionNodeInOperator *inOp = dynamic_cast<QgsExpressionNodeInOperator *>( node ) )
+      {
+        if ( inOp->node()->nodeType() != QgsExpressionNode::ntColumnRef )
+        {
+          return false;
+        }
+
+        const QString fieldName = inOp->node()->dump();
+
+        // Check if all nodes are literals
+        const auto nodes = inOp->list()->list();
+        for ( const auto &valueNode : std::as_const( nodes ) )
+        {
+          if ( valueNode->nodeType() != QgsExpressionNode::ntLiteral )
+          {
+            return false;
+          }
+        }
+
+        if ( !orValuesMap.contains( fieldName ) )
+        {
+          orFieldNames.append( fieldName );
+          orValuesMap.insert( fieldName, *inOp->list()->clone() );
+        }
+        else
+        {
+          for ( const auto &valueNode : std::as_const( nodes ) )
+          {
+            orValuesMap[fieldName].append( valueNode->clone() );
+          }
+        }
+
+        return true;
+      }
+      return false;
+    };
+
+
+    // Second step: build the OR chain of IN operators
+    if ( visitOrNodes( this ) && ! orValuesMap.empty() )
+    {
+
+      // Recursively build the OR chain of IN operators
+      std::function<QgsExpressionNode* ( QList<QString>::const_iterator )> buildOrChain = [&buildOrChain, &orValuesMap, &orFieldNames]( QList<QString>::const_iterator fieldNameit ) -> QgsExpressionNode *
+      {
+
+        QgsExpressionNode *currentNode;
+        auto orValuesIt = orValuesMap.find( *fieldNameit );
+
+        if ( orValuesIt.value().count() == 1 )
+        {
+          currentNode = new QgsExpressionNodeBinaryOperator( boEQ, new QgsExpressionNodeColumnRef( orValuesIt.key() ), orValuesIt.value().at( 0 )->clone() );
+        }
+        else
+        {
+          currentNode = new QgsExpressionNodeInOperator( new QgsExpressionNodeColumnRef( orValuesIt.key() ), orValuesIt.value().clone() );
+        }
+
+        if ( fieldNameit + 1 == orFieldNames.cend() )
+        {
+          return currentNode;
+        }
+        else
+        {
+          QgsExpressionNode *nextNode = buildOrChain( fieldNameit + 1 );
+          return new QgsExpressionNodeBinaryOperator( boOr, currentNode, nextNode );
+        }
+      };
```

Comment:
```suggestion

      QgsExpressionNode *currentNode = nullptr;
      for ( auto it :  orValuesMap )
      {
       QgsExpressionNode *node;
           if ( it->.count() == 1 )
        {
          currentNode = new QgsExpressionNodeBinaryOperator( boEQ, new QgsExpressionNodeColumnRef( it.key() ), it->at( 0 )->clone() );
        }
        else
        {
          currentNode = new QgsExpressionNodeInOperator( new QgsExpressionNodeColumnRef( it.key() ), it->clone() );
        }

        if ( !currentNode )
           currentNode = node;
        else
        {
          currentNode = new QgsExpressionNodeBinaryOperator( boOr, currentNode, node );
        }
      };
       mCompiledSimplifiedNode.reset( currentNode );
```

Can we not replace recursion here with something approximatively like that? and use unique_ptr in the whole process

Reply:
Done, no recursion now and a new handful of unique ptr.